pkg/metrics/metrics.go

```diff
@@ -0,0 +1,129 @@
+package metrics
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	upStatus = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "up_status",
+		Help: "The up status of the build-controller.",
+	})
+
+	buildCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "build_count",
+			Help: "Build creation count.",
+		},
+		[]string{"build"})
+
+	buildRunCount = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_count",
+			Help: "BuildRun execution count.",
+		},
+		[]string{"buildrun"})
+
+	buildRunEstablishDuration = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_establish_duration",
+			Help: "BuildRun establish duration in seconds.",
+		},
+		[]string{"buildrun"})
+
+	buildRunReadyDuration = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "buildrun_ready_duration",
+			Help: "BuildRun ready duration in seconds.",
+		},
+		[]string{"buildrun"})
+
+	buildRunReadyDurationHis = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "buildrun_ready_duration_his",
+			Help:    "BuildRun ready duration histogram in seconds.",
+			Buckets: prometheus.LinearBuckets(50, 50, 10),
+		},
+		[]string{"buildrun"})
+)
+
+//CustomBuildMetrics This struct implements the Collector interface to be registered in the middleware
+//by the Prometheus server
+type CustomBuildMetrics struct {
+	UpStatus                      prometheus.Gauge
+	BuildCount                    prometheus.GaugeVec
+	BuildRunCount                 prometheus.GaugeVec
+	BuildRunEstablishDuration     prometheus.GaugeVec
+	BuildRunReadyDuration         prometheus.GaugeVec
+	BuildRunReadyDurationHis      prometheus.HistogramVec
+}
+
+//ProxyMetrics of the package level item.
+var BuildMetrics = CustomBuildMetrics{
+	UpStatus:                         upStatus,
+	BuildCount:                       *buildCount,
+	BuildRunCount:                    *buildRunCount,
+	BuildRunEstablishDuration:        *buildRunEstablishDuration,
+	BuildRunReadyDuration:            *buildRunReadyDuration,
+	BuildRunReadyDurationHis:         *buildRunReadyDurationHis,
+}
+
+//Describe Implementation of the Collector Describe interface function
+func (r CustomBuildMetrics) Describe(descCh chan<- *prometheus.Desc) {
+	r.UpStatus.Describe(descCh)
+	r.BuildCount.Describe(descCh)
+	r.BuildRunCount.Describe(descCh)
+	r.BuildRunEstablishDuration.Describe(descCh)
+	r.BuildRunReadyDuration.Describe(descCh)
+	r.BuildRunReadyDurationHis.Describe(descCh)
+}
+
+//Collect Implementation of the Collector Collect interface function
+func (r CustomBuildMetrics) Collect(collectCh chan<- prometheus.Metric) {
```

Comment:
for what is this func? and who uses it?

Reply:
It should be used in future for some additional metric operations, I copied from example. But OK let me remove it first.